### PR TITLE
[logging] Integration tests

### PR DESCRIPTION
**Pull request by paramite**
This patch sets structure for integration tests starting with logging
plugins of sg-core

Changelog:
 - enables unit tests for loki application
 - refactors unit tests execution to centos image to enable spawning
   loki container correctly
 - introduction of integration tests structure
 - itegration tests for logs handler and elasticsearch and loki applications

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (pleimer) [inline .github/workflows/tests.yml:49]: Curious why do you run the Loki service here?

**Review by pleimer — APPROVED** [review]
LGTM once the issue with the unit tests is resolved

File: .github/workflows/integration.yml
Comment (leifmadsen) [inline .github/workflows/integration.yml:22]: We can bump this to 7.10.2 now that we're using that in STF

File: .github/workflows/integration.yml
Comment (leifmadsen) [inline .github/workflows/integration.yml:13]: Can this be moved to ubi8:latest instead? Otherwise we should consider moving to centos stream

File: .github/workflows/tests.yml
Comment (leifmadsen) [inline .github/workflows/tests.yml:5]: For version tags, might be ideal to put these all in one spot and refer to them via env vars just to make it easier to update this in the future.

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [inline ci/integration/logging/run_sg.sh:8]: It would be better if we could avoid using epel-release and base this on ubi8 which is going to better align everything downstream for us. I recently migrated a couple of the other repositories to ubi8 and there should be some yum repositories loaded that will help cover things that don't exist in native ubi8.

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [inline ci/integration/logging/run_sg.sh:20]: Cloning and building this each time seems a bit heavy no? Is there a dependency here where you need to build sg-bridge? If there is a cross-project interface here, I would like to see it try and clone the same-named branch like we do for STF CI stuff, with a fallback to `master` branch.

File: ci/integration/logging/sg_config.yaml
Comment (leifmadsen) [inline ci/integration/logging/sg_config.yaml:30]:
```suggestion
      connection: http://127.0.0.1:3100
      batchsize: 3
      maxwaittime: 1
```

**Comment by vyzigold** [comment]
It looks like adding data.DEBUG and translating severity from number to string broke most of the log/event unit tests, but otherwise the PR looks ok to me.

**Comment by paramite** [comment]
Yes indeed. This fact I missed and working on fixing it :)

**Comment by paramite** [comment]
Ok unit tests are fine now, but coveralls report seems missing token. Investigating.

**Comment by paramite** [comment]
I wasn't able to move the coverage reporting to unit test script and then realize that I'm actually trying to reinvent the wheel. It's fixed now as you can check at [1]. Unit tests fail because on PR from fork coveralls token is not passed. Going to work on comments now. 

[1] https://github.com/infrawatch/sg-core/actions

**Comment by paramite** [comment]
If I did not missed anything I think this PR is good for next round of review and eventual merge.

File: .github/workflows/integration.yml
Comment (leifmadsen) [reply .github/workflows/integration.yml]: I'd have preferred ubi8 here since that's what we're going to be building against downstream, but since I called out stream I guess this is ok.

File: .github/workflows/integration.yml
Comment (leifmadsen) [inline .github/workflows/integration.yml:23]: Move this in to an ENV like other images.

File: .github/workflows/tests.yml
Comment (leifmadsen) [reply .github/workflows/tests.yml:5]: Disregard this one. Looks fine.

File: .github/workflows/tests.yml
Comment (leifmadsen) [inline .github/workflows/tests.yml:39]: Move this to an ENV like other images.

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [reply ci/integration/logging/run_sg.sh:8]: Yea check this out in smart-gateway repo where I use opstools.repo for qpid-proton-c-devel.

https://github.com/infrawatch/smart-gateway/blob/master/build/test-framework/run_tests.sh

All the other repos we have are on ubi8, and our base image downstream is ubi8, so I'm kind of wanting it all to be in sync as much as possible so we don't get surprised in testing or builds when importing downstream.

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [reply ci/integration/logging/run_sg.sh:21]: I think the simplest thing is to basically check if the same-named branch exists on the sg-bridge side (for coordinated testing if changes on sg-bridge are required to satisfy a change here), and if not, then just pull from quay.io/infrawatch/sg-bridge:latest

It's not that big of a deal, but it could certainly be a bit faster test, although I suspect sg-bridge doesn't ultimately take all that long to deploy.

If you want to make the changes that's fine, if not, I won't block it.

File: ci/integration/logging/run_rsyslog.sh
Comment (leifmadsen) [inline ci/integration/logging/run_rsyslog.sh:21]: I think both ubi8, centos:streams(?) and Fedora all use `dnf` now actually.

**Review by leifmadsen — APPROVED** [review]
Could small comments, but otherwise this looks good to me. Thanks!

File: .github/workflows/integration.yml
Comment (paramite) [reply .github/workflows/integration.yml:23]: Ah I see, ok will do.

File: .github/workflows/tests.yml
Comment (paramite) [reply .github/workflows/tests.yml:39]: Will do.

File: ci/integration/logging/run_sg.sh
Comment (paramite) [reply ci/integration/logging/run_sg.sh:8]: Oh I was not aware that we have this dependency in opstools repo. In that case I can move to ubi8. Thanks for a pointer.

File: ci/integration/logging/run_sg.sh
Comment (paramite) [reply ci/integration/logging/run_sg.sh:21]: So ... you want me to run sg-bridge in separate container instead of sg-core container and in case the sg-bridge repo does not have the same topic branch use the prebuilt image ... got it.

File: ci/integration/logging/run_rsyslog.sh
Comment (paramite) [reply ci/integration/logging/run_rsyslog.sh:21]: Old habits die hard :)

**Comment by paramite** [comment]
@leifmadsen I hope that you will be absolutely happy now :)

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [inline ci/integration/logging/run_sg.sh:23]: Do we need a test at some point that if `go mod tidy` actually results in changes, that we should fail on that?

File: ci/integration/logging/run_validation.sh
Comment (leifmadsen) [inline ci/integration/logging/run_validation.sh:20]:
```suggestion
#TODO(mmagr): adapt elasticsearch plugin to create index templates avoiding unnecessary prefix and suffix
```

File: ci/integration/logging/run_validation.sh
Comment (leifmadsen) [inline ci/integration/logging/run_validation.sh:34]: Would it just make more sense to use single-quotes for the query so you don't have to escape the double-quotes? Maybe there is something going on here I don't understand though.

File: ci/unit/run_tests.sh
Comment (leifmadsen) [inline ci/unit/run_tests.sh:18]:
```suggestion
dnf install -y git golang gcc make glibc-langpack-en qpid-proton-c-devel
```

**Review by leifmadsen — APPROVED** [review]

**Comment by leifmadsen** [comment]
> @leifmadsen I hope that you will be absolutely happy now :)

Very close lol

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [reply ci/integration/logging/run_sg.sh:23]: PS not for fixing now, just asking. If you think so, we can just open an issue to add it as an enhancement.

Comment (paramite) [reply ci/integration/logging/run_sg.sh:23]: Maybe. The reason of tidy is that for some reason without it build script was failing on missing dependencies, because the legacy go path mode was being used and the dependencies were not installed in go path. SO we definitely need some improvements on that side.

File: ci/integration/logging/run_validation.sh
Comment (paramite) [reply ci/integration/logging/run_validation.sh:34]: That is a left over when I was trying to query log records not only by phrase, but also by host (saved in env variable) and by beginning of timestamp (also saved in env var). But I failed miserably on that front. Since I had al the queries written with double quotes double quotes. But yeah I'll fix it.

File: ci/integration/logging/run_sg.sh
Comment (leifmadsen) [inline ci/integration/logging/run_sg.sh:8]: Huh I learned a new trick today that if you surround the EOF in single quotes like `'EOF'` it won't expand the `$` etc.

```
cat > /tmp/foo <<'EOF'
> $(bar) foo
> EOF

cat /tmp/foo 
$(bar) foo
```

**Comment by leifmadsen** [comment]
I saw one of the tests failed, so I re-ran it just in case. I don't want to keep nit-picking at this. Looks good. As soon as all the tests pass feel free to merge this. Thanks!

**Comment by paramite** [comment]
> I saw one of the tests failed, so I re-ran it just in case. I don't want to keep nit-picking at this. Looks good. As soon as all the tests pass feel free to merge this. Thanks!

Yes one of the metrics unit test is unstable and sometimes fail. We will have to figure this out in another patch.

**Pull request merged**
